node/network/collator-protocol/src/lib.rs

```diff
@@ -71,39 +86,39 @@ impl CollatorProtocolSubsystem {
 	/// Start the collator protocol.
 	/// If `id` is `Some` this is a collator side of the protocol.
 	/// If `id` is `None` this is a validator side of the protocol. 
-	pub fn new(id: Option<CollatorId>) -> Self {
+	/// Caller must provide a registry for prometheus metrics.
+	pub fn new(id: Option<CollatorId>, registry: &prometheus::Registry)
```

Comment:
I think `&Registry` is not always available, `real_overseer` accepts `Option<&Registry>`. we could do the same and call `register` function instead of `try_register`.

Wouldn't it be better to make `ProtocolSide` public and pass it here instead?

`ProtocolSide` could have two constructors:
```rust
impl ProtocolSide {
    pub fn validator(registry: Option<&Registry>) -> Result<Self> {
        ...
    }
    pub fn collator(id: CollatorId, registry: Option<&Registry>) -> Result<Self> {
        ...
    }
}
```

Reply:
In the original PR [there were grumbles agains it](https://github.com/paritytech/polkadot/pull/1659#discussion_r484519472)

Reply:
But how was it resolved? We still are making a choice at startup depending on whether `id` is `None` or not?

Reply:
Yep, I see no other options